core/src/main/scala/com/databricks/labs/remorph/parsers/ParserCommon.scala

```diff
@@ -1,23 +1,100 @@
 package com.databricks.labs.remorph.parsers
 
-import org.antlr.v4.runtime.{RuleContext, ParserRuleContext}
+import com.typesafe.scalalogging.LazyLogging
 import org.antlr.v4.runtime.misc.Interval
-import org.antlr.v4.runtime.tree.{AbstractParseTreeVisitor, ParseTree}
+import org.antlr.v4.runtime.tree.{AbstractParseTreeVisitor, ParseTree, ParseTreeVisitor, RuleNode}
+import org.antlr.v4.runtime.{ParserRuleContext, RuleContext}
 
 import scala.collection.JavaConverters._
-trait ParserCommon[A] { self: AbstractParseTreeVisitor[A] =>
+
+trait ParserCommon[A] extends ParseTreeVisitor[A] with LazyLogging { self: AbstractParseTreeVisitor[A] =>
   protected def occursBefore(a: ParseTree, b: ParseTree): Boolean = {
     a != null && b != null && a.getSourceInterval.startsBeforeDisjoint(b.getSourceInterval)
   }
 
   def visitMany[R <: RuleContext](contexts: java.lang.Iterable[R]): Seq[A] = contexts.asScala.map(_.accept(self)).toSeq
 
   /**
-   * Creates a string representation of the text represented by the given ANTLR ParserRuleContext.
+   * <p>
+   *   An implementation sof this should return some type of ir.UnresolvedXYZ object that represents the
+   *   unresolved input that we have no visitor for. This is used in the default visitor to wrap the
+   *   unresolved input.
+   * </p>
+   * @param msg What message the unresolved object should contain
+   * @return An instance of the type returned by the implementing visitor
+   */
+  protected def unresolved(msg: String): A
+
+  protected override def defaultResult(): A = {
+    unresolved(
+      s"Unimplemented visitor $caller in class $implementor" +
+        s" for ${getTextFromParserRuleContext(currentNode.getRuleContext)}")
+  }
+
+  /**
+   * <p>
+   *   Creates a string representation of the text represented by the given ANTLR ParserRuleContext.
+   * </p>
+   * <p>
+   *   Note that this should exactly reflect the original input text as bounded by the source interval
+   *   recorded by the parser.
+   * </p>
+   */
+  def getTextFromParserRuleContext(ctx: RuleContext): String = ctx match {
```

Comment:
nitpicking:

```suggestion
  protected def getTextFromParserRuleContext(ctx: RuleContext): String = ctx match {
```

Reply:
It is accessed from outside the visitor in tests